### PR TITLE
Enforce that the optimizer closure is executed when `optimizer_step` is overridden

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of #9349

This has the advantage that the `ClosureResult` is no longer `Optional`.

### Does your PR introduce any breaking changes? If yes, please list them.

`optimizer_step` now requires that the closure is executed. Not doing it would skip the entire `training_step`, `zero_grad`, and `backward`. We consider this to be a bug so it's better to explicitly fail. Not doing it could also break progress tracking and fault tolerance.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/optimizer/optimizer_loop.py [hunk 1]
```diff
@@ -158,16 +158,15 @@ def _run_optimization(
         else:
             self._optimizer_step(optimizer, opt_idx, batch_idx, closure)
 
-        result = closure.get_result()
+        result = closure.result
 
-        if result:
+        if result.loss:
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimizer/optimizer_loop.py:163]: this should be a None check. The loss is a tensor here 

```suggestion
        if result.loss is not None:
```

File: pytorch_lightning/loops/closure.py [hunk 2]
```diff
@@ -53,25 +54,29 @@ def __init__(self) -> None:
         super().__init__()
         self._result: Optional[ClosureResult] = None
 
-    def get_result(self) -> Optional[ClosureResult]:
```
Comment (awaelchli) [inline pytorch_lightning/loops/closure.py:56]: the naming here was intentional. We are only allowed to access the result once. In that sense, it really is not a property. I vote for keeping this a method.

File: pytorch_lightning/loops/closure.py [hunk 3]
```diff
@@ -53,25 +54,29 @@ def __init__(self) -> None:
...
 
         Once accessed, the internal reference gets reset and the consumer will have to hold on to the reference as long
         as necessary.
         """
-        result = self._result
-        self._result = None  # free memory
+        if self._result is None:
+            raise MisconfigurationException(
+                "The closure hasn't been executed."
+                " HINT: did you call `optimizer_closure()` in your `optimizer_step` hook?"
```
Comment (awaelchli) [inline pytorch_lightning/loops/closure.py:67]: maybe the hint should suggest that the optimizer calls the closure? If an optimizer didn't do this, then an error like this would also be very helpful right?

File: pytorch_lightning/loops/closure.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/closure.py:56]: Sounds good, changed `get_result` to `consume_result` to make it more explicit

File: tests/loops/test_closure.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import pytest
```
Comment (awaelchli) [inline tests/loops/test_closure.py:1]: license :)

File: pytorch_lightning/loops/closure.py [hunk 5]
```diff
@@ -53,25 +54,29 @@ def __init__(self) -> None:
...
         Once accessed, the internal reference gets reset and the consumer will have to hold on to the reference as long
         as necessary.
         """
-        result = self._result
-        self._result = None  # free memory
+        if self._result is None:
+            raise MisconfigurationException(
+                "The closure hasn't been executed."
+                " HINT: did you call `optimizer_closure()` in your `optimizer_step` hook? It could also happen because"
+                " the `optimizer.step(optimizer_closure)` call did not execute it internally"
```
Comment (awaelchli) [inline pytorch_lightning/loops/closure.py:67]:
```suggestion
                " the `optimizer.step(optimizer_closure)` call did not execute it internally."
```

**Review by awaelchli — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/closure.py [hunk 6]
```diff
@@ -53,25 +54,29 @@ def __init__(self) -> None:
...
 
         Once accessed, the internal reference gets reset and the consumer will have to hold on to the reference as long
         as necessary.
         """
-        result = self._result
-        self._result = None  # free memory
+        if self._result is None:
+            raise MisconfigurationException(
+                "The closure hasn't been executed."
+                " HINT: did you call `optimizer_closure()` in your `optimizer_step` hook? It could also happen because"
```
Comment (tchaton) [inline pytorch_lightning/loops/closure.py:66]: Would be good to be more consistent on the closure name across the codebase, lambda_closure, optimizer_closure, closure, etc..

File: tests/loops/test_closure.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        ):
+            # does not call `optimizer_closure()`
+            pass
+
+    model = TestModel()
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
+    with pytest.raises(MisconfigurationException, match="The closure hasn't been executed"):
+        trainer.fit(model)
+
+    class TestModel(BoringModel):
```
Comment (tchaton) [inline tests/loops/test_closure.py:35]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by daniellepintz** [comment]
@carmocca it looks like this PR introduced a test failure in `tests/utilities/test_parsing.py::test_is_picklable`

https://github.com/PyTorchLightning/pytorch-lightning/runs/3543485388

**Comment by carmocca** [comment]
> @carmocca it looks like this PR introduced a test failure in `tests/utilities/test_parsing.py::test_is_picklable`

I can't think of a reason why the changes in this PR would make that test fail. Since 1.10 testing comes from nightly, it might be caused by a change in PyTorch and this was the first PR to see it.

File: pytorch_lightning/loops/closure.py [hunk 8]
```diff
@@ -53,25 +54,29 @@ def __init__(self) -> None:
...
+    def consume_result(self) -> ClosureResult:
         """The cached result from the last time the closure was called.
 
         Once accessed, the internal reference gets reset and the consumer will have to hold on to the reference as long
         as necessary.
         """
-        result = self._result
-        self._result = None  # free memory
+        if self._result is None:
+            raise MisconfigurationException(
```
Comment (awaelchli) [inline pytorch_lightning/loops/closure.py:64]: @carmocca FB had a problem here where they got the error message but not calling the closure was actually intentional. Are we ok with downgrading this to a warning message?

Comment (carmocca) [reply pytorch_lightning/loops/closure.py:64]: We can downgrade this to a warning if necessary, however, if there isn't a reason for allowing this, I still believe this should be an error.

We can make this a warning and a deprecation and then convert this back to an error two minor releases later.

@ananthsub what do you think of this solution? Do you have an argument for why skipping the closure should be allowed?
